ffi/SDL2_0.lua

```diff
@@ -293,64 +348,74 @@ function S.waitForEvent(sec, usec)
+        else
+            if pointers[0] and pointers[0].down then
+                slot = 0 -- left mouse button down
+                genTouchMoveEvent(event, slot, event.motion.x, event.motion.y)
+            end
+            if pointers[1] and pointers[1].down then
+                slot = 1 -- right mouse button down
+                genTouchMoveEvent(event, slot, event.motion.x, event.motion.y)
             end
-            genEmuEvent(C.EV_SYN, C.SYN_REPORT, 0)
         end
-    elseif event.type == SDL.SDL_MOUSEBUTTONUP
+    elseif event.type == SDL.SDL_MOUSEBUTTONUP and event.button.which ~= SDL_TOUCH_MOUSEID
         or event.type == SDL.SDL_FINGERUP then
-        is_in_touch = false
-        genEmuEvent(C.EV_ABS, C.ABS_MT_TRACKING_ID, -1)
-        genEmuEvent(C.EV_SYN, C.SYN_REPORT, 0)
-    elseif event.type == SDL.SDL_MOUSEBUTTONDOWN
+        local is_finger = event.type == SDL.SDL_FINGERUP
+        local slot
+        if is_finger then
+            slot = getFingerSlot(event)
+        elseif event.button.button == SDL_BUTTON_RIGHT then
+            slot = 1
+        else -- SDL_BUTTON_LEFT
+            slot = 0
+        end
+
+        local x = is_finger and event.tfinger.x * S.w or event.button.x
+        local y = is_finger and event.tfinger.y * S.h or event.button.y
+        genTouchUpEvent(event, slot, x, y)
+        if is_finger then
+            pointers[tonumber(event.tfinger.fingerId)] = nil
+        else
+            pointers[slot] = nil
+        end
+    elseif event.type == SDL.SDL_MOUSEBUTTONDOWN and event.button.which ~= SDL_TOUCH_MOUSEID
         or event.type == SDL.SDL_FINGERDOWN then
         local is_finger = event.type == SDL.SDL_FINGERDOWN
-        if not is_finger and event.button.button ~= SDL_BUTTON_LEFT then
-            -- Not a left-click?
+        if not is_finger and not (event.button.button == SDL_BUTTON_LEFT or event.button.button == SDL_BUTTON_RIGHT) then
+            -- We don't do anything with extra buttons for now.
             return false, C.ENOSYS
         end
+
         -- use mouse click to simulate single tap
-        is_in_touch = true
-        genEmuEvent(C.EV_ABS, C.ABS_MT_TRACKING_ID, 0)
-        genEmuEvent(C.EV_ABS, C.ABS_MT_POSITION_X,
-            is_finger and event.tfinger.x * S.w or event.button.x)
-        genEmuEvent(C.EV_ABS, C.ABS_MT_POSITION_Y,
-            is_finger and event.tfinger.y * S.h or event.button.y)
-        genEmuEvent(C.EV_SYN, C.SYN_REPORT, 0)
+        local slot
+        if is_finger then
+            slot = getFingerSlot(event)
+        elseif event.button.button == SDL_BUTTON_RIGHT then
+            slot = 1
+        else -- SDL_BUTTON_LEFT
+            slot = 0
+        end
+        local x = is_finger and event.tfinger.x * S.w or event.button.x
+        local y = is_finger and event.tfinger.y * S.h or event.button.y
+        setPointerDownState(slot, true)
+        genTouchDownEvent(event, slot, x, y)
```

Comment:
This may be called for !is_finger, while it *will* access `event.tfinger.fingerId`, I'm not sure what it points to in the SDL magic union in those cases ;).

Reply:
Oh right, a minor oversight for the mouse, though it doesn't really matter in practice.

Reply:
The most obvious would probably be:
```diff
diff --git a/ffi/SDL2_0.lua b/ffi/SDL2_0.lua
index 4823dcac..ee5a767d 100644
--- a/ffi/SDL2_0.lua
+++ b/ffi/SDL2_0.lua
@@ -235,8 +235,9 @@ local function getFingerSlot(event)
 end
 
 local function genTouchDownEvent(event, slot, x, y)
+    local is_finger = event.type == SDL.SDL_FINGERDOWN
     genEmuEvent(C.EV_ABS, C.ABS_MT_SLOT, slot)
-    genEmuEvent(C.EV_ABS, C.ABS_MT_TRACKING_ID, tonumber(event.tfinger.fingerId))
+    genEmuEvent(C.EV_ABS, C.ABS_MT_TRACKING_ID, is_finger and tonumber(event.tfinger.fingerId) or slot)
     genEmuEvent(C.EV_ABS, C.ABS_MT_POSITION_X, x)
     genEmuEvent(C.EV_ABS, C.ABS_MT_POSITION_Y, y)
     genEmuEvent(C.EV_SYN, C.SYN_REPORT, 0)
```

I prefer tracking id to not be constantly 0 and 1 'cause if nothing else it's clearer to read. :sweat_smile: